File: app/src/main/java/me/mitul/aij/utils/JustifiedTextView.java
package me.mitul.aij.utils;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint.Align;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.ViewTreeObserver.OnGlobalLayoutListener;

import java.util.ArrayList;
import java.util.List;

public class JustifiedTextView extends View {
    private boolean hasTextBeenDrown = false;
    private Context context;
    private TextPaint textPaint;
    private int lineSpace = 0;
    private int lineHeight = 0;
    private int textAreaWidth = 0;
    private int measuredViewHeight = 0;
    private int measuredViewWidth = 0;
    private String text;
    private List<String> lineList = new ArrayList<>();

    public JustifiedTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        constructor(context, attrs);
    }

    public JustifiedTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        constructor(context, attrs);
    }

    public JustifiedTextView(Context context) {
        super(context);
        constructor(context, null);
    }

    private void constructor(Context context, AttributeSet attrs) {
        this.context = context;
        XmlToClassAttribHandler xmlParser = new XmlToClassAttribHandler(this.context, attrs);
        initTextPaint();

        if (attrs != null) {
            String text;
            int textColor;
            int textSize;
            int textSizeUnit;
            text = xmlParser.getTextValue();
            textColor = xmlParser.getColorValue();
            textSize = xmlParser.getTextSize();
            textSizeUnit = xmlParser.gettextSizeUnit();
            setText(text);
            setTextColor(textColor);
            if (textSizeUnit == -1) setTextSize(textSize);
            else setTextSize(textSizeUnit, textSize);
//			setText(XmlToClassAttribHandler.GetAttributeStringValue(mContext, attrs, namespace, key, ""));
            setTypeFace(xmlParser.getTypeFace());
            //setTypeFace(Typeface.create("serif", Typeface.NORMAL));
        }

        ViewTreeObserver observer = getViewTreeObserver();
        observer.addOnGlobalLayoutListener(new OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                if (hasTextBeenDrown) return;
                hasTextBeenDrown = true;
                setTextAreaWidth(getWidth() - (getPaddingLeft() + getPaddingRight()));
                calculate();
            }
        });
    }

    private void calculate() {
        setLineHeight(getTextPaint());
        lineList.clear();
        lineList = divideOriginalTextToStringLineList(getText());
        setMeasuredDimentions(lineList.size(), getLineHeight(), getLineSpace());
        measure(getMeasuredViewWidth(), getMeasuredViewHeight());
    }

    private void initTextPaint() {
        textPaint = new TextPaint(TextPaint.ANTI_ALIAS_FLAG);
        textPaint.setTextAlign(Align.LEFT);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (getMeasuredViewWidth() > 0) {
            requestLayout();
            setMeasuredDimension(getMeasuredViewWidth(), getMeasuredViewHeight());
        } else super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        int rowIndex = getPaddingTop();
        int colIndex = 0;
        if (getAlignment() == Align.RIGHT) colIndex = getPaddingLeft() + getTextAreaWidth();
        else colIndex = getPaddingLeft();
        for (int i = 0; i < lineList.size(); i++) {
            rowIndex += getLineHeight() + getLineSpace();
            canvas.drawText(lineList.get(i), colIndex, rowIndex, getTextPaint());
        }
    }

    private List<String> divideOriginalTextToStringLineList(String originalText) {
        ArrayList<String> listStringLine = new ArrayList<String>();
        String line = "";
        float textWidth;
        String[] paragraphList = originalText.split("\n");
        for (String paragraph : paragraphList) {
            String[] wordList = paragraph.split(" ");
            for (int i = 0; i < wordList.length; i++) {
                line += wordList[i] + " ";
                textWidth = getTextPaint().measureText(line);
                if (getTextAreaWidth() == textWidth) {
                    listStringLine.add(line);
                    line = "";//make line clear
                    continue;
                } else if (getTextAreaWidth() < textWidth) {
                    int lastWordCount = wordList[i].length();
                    line = line.substring(0, line.length() - lastWordCount - 1);
                    if (line.trim().length() == 0)
                        continue;
                    line = justifyTextLine(textPaint, line.trim(), getTextAreaWidth());
                    listStringLine.add(line);
                    line = "";
                    i++;
                    continue;
                }

                if (i == wordList.length - 1) {
                    listStringLine.add(line);
                    line = "";
                }
            }
        }
        return listStringLine;
    }

    private String justifyTextLine(TextPaint textPaint, String lineString, int textAreaWidth) {
        int gapIndex = 0;
        float lineWidth = textPaint.measureText(lineString);
        while (lineWidth < textAreaWidth && lineWidth > 0) {
            gapIndex = lineString.indexOf(" ", gapIndex + 2);
            if (gapIndex == -1) {
                gapIndex = 0;
                gapIndex = lineString.indexOf(" ", gapIndex + 1);
                if (gapIndex == -1) return lineString;
            }
            lineString = lineString.substring(0, gapIndex) + "  " + lineString.substring(gapIndex + 1);
            lineWidth = textPaint.measureText(lineString);
        }
        return lineString;
    }

    private void setLineHeight(TextPaint textPaint) {
        Rect bounds = new Rect();
        String sampleStr = "bla bla, who is the crazy world of all";
        textPaint.getTextBounds(sampleStr, 0, sampleStr.length(), bounds);
        setLineHeight(bounds.height());
    }

    public void setMeasuredDimentions(int lineListSize, int lineHeigth, int lineSpace) {
        int height = lineListSize * (lineHeigth + lineSpace) + lineSpace;
        height += getPaddingRight() + getPaddingLeft();
        setMeasuredViewHeight(height);
        setMeasuredViewWidth(getWidth());
    }

    private int getTextAreaWidth() {
        return textAreaWidth;
    }

    private void setTextAreaWidth(int textAreaWidth) {
        this.textAreaWidth = textAreaWidth;
    }

    private int getLineHeight() {
        return lineHeight;
    }

    private void setLineHeight(int lineHeight) {
        this.lineHeight = lineHeight;
    }

    private int getMeasuredViewHeight() {
        return measuredViewHeight;
    }

    private void setMeasuredViewHeight(int measuredViewHeight) {
        this.measuredViewHeight = measuredViewHeight;
    }

    private int getMeasuredViewWidth() {
        return measuredViewWidth;
    }

    private void setMeasuredViewWidth(int measuredViewWidth) {
        this.measuredViewWidth = measuredViewWidth;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
        calculate();
        invalidate();
    }

    public void setText(int resId) {
        setText(context.getResources().getString(resId));
    }

    public Typeface getTypeFace() {
        return getTextPaint().getTypeface();
    }

    public void setTypeFace(Typeface typeFace) {
        getTextPaint().setTypeface(typeFace);
    }

    public float getTextSize() {
        return getTextPaint().getTextSize();
    }

    private void setTextSize(float textSize) {
        getTextPaint().setTextSize(textSize);
        calculate();
        invalidate();
    }

    public void setTextSize(int unit, float textSize) {
        textSize = TypedValue.applyDimension(unit, textSize, context.getResources().getDisplayMetrics());
        setTextSize(textSize);
    }

    public TextPaint getTextPaint() {
        return textPaint;
    }

    public void setTextPaint(TextPaint textPaint) {
        this.textPaint = textPaint;
    }

    public void setLineSpacing(int lineSpace) {
        this.lineSpace = lineSpace;
        invalidate();
    }

    public int getTextColor() {
        return getTextPaint().getColor();
    }

    public void setTextColor(int textColor) {
        getTextPaint().setColor(textColor);
        invalidate();
    }

    public int getLineSpace() {
        return lineSpace;
    }

    public Align getAlignment() {
        return getTextPaint().getTextAlign();
    }

    public void setAlignment(Align align) {
        getTextPaint().setTextAlign(align);
        invalidate();
    }
}

File: app/src/main/java/me/mitul/aij/aij/JustifiedTextView.kt
package me.mitul.aij.aij

import android.content.Context
import android.content.res.Resources
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint.Align
import android.graphics.Rect
import android.graphics.Typeface
import android.text.TextPaint
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewTreeObserver.OnGlobalLayoutListener

class JustifiedTextView : View {
    private var hasTextBeenDrown = false
    private var context: Context? = null
    private var textPaint: TextPaint? = null
    private var lineSpace = 0
    private var lineHeight = 0
    private var textAreaWidth = 0
    private var measuredViewHeight = 0
    private var measuredViewWidth = 0

    var text: String? = null
        set(text) {
            field = text
            calculate()
            invalidate()
        }

    private var lineList: MutableList<String> = ArrayList()

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) :
            super(context, attrs, defStyle) {
        constructor(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        constructor(context, attrs)
    }

    constructor(context: Context) : super(context) {
        constructor(context, null)
    }

    private fun constructor(context: Context, attrs: AttributeSet?) {
        this.context = context
        val xmlParser = XmlToClassAttribHandler(this.context, attrs)
        initTextPaint()
        if (attrs != null) {
            val text: String = xmlParser.textValue
            val textSize: Int = xmlParser.textSize
            val textColor: Int = xmlParser.colorValue
            val textSizeUnit: Int = xmlParser.gettextSizeUnit()
            this.text = text
            this.textColor = textColor
            if (textSizeUnit == -1) this.textSize = textSize.toFloat()
            else setTextSize(textSizeUnit, textSize.toFloat())
            typeFace = xmlParser.typeFace
            //setTypeFace(Typeface.create("serif", Typeface.NORMAL));
        }
        val observer = getViewTreeObserver()
        observer.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (hasTextBeenDrown) return
                hasTextBeenDrown = true
                textAreaWidth = width - (getPaddingLeft() + getPaddingRight())
                calculate()
            }
        })
    }

    private fun calculate() {
        setLineHeight(textPaint)
        lineList.clear()
        lineList = divideOriginalTextToStringLineList(text)
        setMeasuredDimentions(lineList.size, lineHeight, lineSpace)
        measure(measuredViewWidth, measuredViewHeight)
    }

    private fun initTextPaint() {
        textPaint = TextPaint(TextPaint.ANTI_ALIAS_FLAG)
        textPaint!!.textAlign = Align.LEFT
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (measuredViewWidth > 0) {
            requestLayout()
            setMeasuredDimension(measuredViewWidth, measuredViewHeight)
        } else super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        var rowIndex = paddingTop
        val colIndex: Int =
            if (alignment == Align.RIGHT) getPaddingLeft() + textAreaWidth
            else getPaddingLeft()
        for (i in lineList.indices) {
            rowIndex += lineHeight + lineSpace
            canvas.drawText(lineList[i], colIndex.toFloat(), rowIndex.toFloat(), textPaint!!)
        }
    }

    private fun divideOriginalTextToStringLineList(originalText: String?): MutableList<String> {
        val listStringLine = ArrayList<String>()
        var line = ""
        var textWidth: Float
        val paragraphList = originalText!!.split("\\n".toRegex())
            .dropLastWhile { it.isEmpty() }.toTypedArray()
        for (paragraph in paragraphList) {
            val wordList = paragraph.split("\\s".toRegex())
                .dropLastWhile { it.isEmpty() }.toTypedArray()
            var i = 0
            while (i < wordList.size) {
                line += wordList[i] + " "
                textWidth = textPaint!!.measureText(line)
                if (textAreaWidth.toFloat() == textWidth) {
                    listStringLine.add(line)
                    line = "" //make line clear
                    i++
                    continue
                } else if (textAreaWidth < textWidth) {
                    val lastWordCount = wordList[i].length
                    line = line.substring(0, line.length - lastWordCount - 1)
                    if (line.trim { it <= ' ' }.isEmpty()) {
                        i++
                        continue
                    }
                    line = justifyTextLine(textPaint, line.trim { it <= ' ' }, textAreaWidth)
                    listStringLine.add(line)
                    line = ""
                    i++
                    i++
                    continue
                }
                if (i == wordList.size - 1) {
                    listStringLine.add(line)
                    line = ""
                }
                i++
            }
        }
        return listStringLine
    }

    private fun justifyTextLine(
        textPaint: TextPaint?,
        lineString: String,
        textAreaWidth: Int,
    ): String {
        var lineString = lineString
        var gapIndex = 0
        var lineWidth = textPaint!!.measureText(lineString)
        while (lineWidth < textAreaWidth && lineWidth > 0) {
            gapIndex = lineString.indexOf(" ", gapIndex + 2)
            if (gapIndex == -1) {
                gapIndex = 0
                gapIndex = lineString.indexOf(" ", gapIndex + 1)
                if (gapIndex == -1) return lineString
            }
            lineString = lineString.substring(0, gapIndex) + "  " +
                    lineString.substring(gapIndex + 1)
            lineWidth = textPaint.measureText(lineString)
        }
        return lineString
    }

    private fun setLineHeight(textPaint: TextPaint?) {
        val bounds = Rect()
        val sampleStr = "bla bla, who is the crazy world of all"
        textPaint!!.getTextBounds(sampleStr, 0, sampleStr.length, bounds)
        setLineHeight(bounds.height())
    }

    fun setMeasuredDimentions(lineListSize: Int, lineHeigth: Int, lineSpace: Int) {
        var height = lineListSize * (lineHeigth + lineSpace) + lineSpace
        height += getPaddingRight() + getPaddingLeft()
        measuredViewHeight = height
        measuredViewWidth = width
    }

    private fun setLineHeight(lineHeight: Int) {
        this.lineHeight = lineHeight
    }

    fun setText(resId: Int) {
        text = context!!.resources.getString(resId)
    }

    var typeFace: Typeface?
        get() = textPaint!!.typeface
        set(typeFace) {
            textPaint!!.setTypeface(typeFace)
        }
    var textSize: Float
        get() = textPaint!!.textSize
        private set(textSize) {
            textPaint!!.textSize = textSize
            calculate()
            invalidate()
        }

    fun setTextSize(unit: Int, textSize: Float) {
        var textSize = textSize
        textSize = TypedValue.applyDimension(unit, textSize, context!!.resources.displayMetrics)
        this.textSize = textSize
    }

    fun setLineSpacing(lineSpace: Int) {
        this.lineSpace = lineSpace
        invalidate()
    }

    var textColor: Int
        get() = textPaint!!.color
        set(textColor) {
            textPaint!!.setColor(textColor)
            invalidate()
        }
    var alignment: Align?
        get() = textPaint!!.textAlign
        set(align) {
            textPaint!!.textAlign = align
            invalidate()
        }
}

internal class XmlToClassAttribHandler(
    private val mContext: Context?,
    private val mAttributeSet: AttributeSet?,
) {
    private val KEY_TEXT_SIZE = "textSize"
    private val mRes: Resources = mContext!!.resources
    private val namespace = "http://me.mitul.aij"

    val textValue: String
        get() {
            val KEY_TEXT = "text"
            var value = mAttributeSet!!.getAttributeValue(namespace, KEY_TEXT) ?: return ""
            if (value.length > 1 && value[0] == '@' && value.contains("@string/")) {
                val resId = mRes.getIdentifier(
                    mContext!!.packageName + ":" + value.substring(1),
                    null,
                    null
                )
                value = mRes.getString(resId)
            }
            return value
        }

    val typeFace: Typeface = Typeface.createFromAsset(mContext!!.assets, "font/f011.ttf")

    val colorValue: Int
        get() {
            val KEY_TEXT_COLOR = "textColor"
            val value = mAttributeSet!!.getAttributeValue(namespace, KEY_TEXT_COLOR)
            var color = Color.BLACK
            if (value == null) return color
            if (value.length > 1 && value[0] == '@' && value.contains("@color/")) {
                val resId = mRes.getIdentifier(
                    mContext!!.packageName + ":" + value.substring(1),
                    null,
                    null
                )
                color = mRes.getColor(resId)
                return color
            }
            color = try {
                Color.parseColor(value)
            } catch (e: Exception) {
                return Color.BLACK
            }
            return color
        }
    val textSize: Int
        get() {
            var textSize = 30
            val value =
                mAttributeSet!!.getAttributeValue(namespace, KEY_TEXT_SIZE) ?: return textSize
            if (value.length > 1 && value[0] == '@' && value.contains("@dimen/")) {
                val resId = mRes.getIdentifier(
                    mContext!!.packageName + ":" + value.substring(1),
                    null,
                    null
                )
                textSize = mRes.getDimensionPixelSize(resId)
                return textSize
            }
            textSize = try {
                value.substring(0, value.length - 2).toInt()
            } catch (e: Exception) {
                return 14
            }
            return textSize
        }

    fun gettextSizeUnit(): Int {
        val value = mAttributeSet!!.getAttributeValue(namespace, KEY_TEXT_SIZE)
            ?: return TypedValue.COMPLEX_UNIT_SP
        try {
            val type = value.substring(value.length - 2)
            when (type) {
                "dp" -> return TypedValue.COMPLEX_UNIT_DIP
                "sp" -> return TypedValue.COMPLEX_UNIT_SP
                "pt" -> return TypedValue.COMPLEX_UNIT_PT
                "mm" -> return TypedValue.COMPLEX_UNIT_MM
                "in" -> return TypedValue.COMPLEX_UNIT_IN
                "px" -> return TypedValue.COMPLEX_UNIT_PX
            }
        } catch (e: Exception) {
            return -1
        }
        return -1
    }
}
